### DISCUSSION TITLE
Create a folder, move file to it and add check list

### DISCUSSION THREAD

**Discussion by carlosbaraza** [SilentVoid13/Templater#569] [Templates Showcase]
Every time I travel, I complete a checklist, and I like to have a folder to to add all my tickets and other resources, so everything is organised.

The following template does:
- Create a `Resources/Travel/YYYY-MM-DD` and `Resources/Travel/YYYY-MM-DD/_resources` folders
- Move the file to the folder and rename to `Travel Checklist`
- Add the template checklist

```
<%*
    // YYYY-MM-DD
    function formatDate(date) {
        let d = new Date(date);
        let month = '' + (d.getMonth() + 1);
        let day = '' + d.getDate();
        let year = d.getFullYear();

        if (month.length < 2) month = '0' + month;
        if (day.length < 2) day = '0' + day;

        return [year, month, day].join('-');
    }

    async function createFolder(dir, times = 1) {
        if (times > 5) {
            throw new Error('createFolder: too many tries');
        }
        try {
            let newDir = dir;
            if (times > 1) newDir = `${dir} (${times})`;
            await this.app.vault.createFolder(newDir);
            return newDir;
        } catch (err) {
            return createFolder(dir, times + 1);
        }
    }

    const today = formatDate(new Date());
    let dir = `Resources/Travel/${today}`;
    dir = await createFolder(dir);
    const resourcesFolder = `${dir}/_resources`;
    await createFolder(resourcesFolder);

    await tp.file.move(`${dir}/Travel Checklist`);
%>
- [ ] Keys to come back
- [ ] Check in
- [ ] Sun cream
- [ ] Glasses

Clothes
- [ ] Underwear
```

**Comment by philoserf** [comment]
I think this can be simplified to:

```markdown
<%*
let dir = "Resources/Travel/" + tp.date.now("YYYY-MM-DD");
await this.app.vault.createFolder(dir);
this.app.vault.createFolder(dir + "/_resources");
tp.file.move(dir + "/Travel Checklist");
%>
- [ ] Keys to come back
- [ ] Check in
- [ ] Sun cream
- [ ] Glasses

Clothes
- [ ] Underwear
```

**Reply by carlosbaraza** [reply]
The `tp.date.now()` is a good trick. However, the `this.app.vault.createFolder(dir)` will fail if the folder already exists. And the `tp.exists` function is only detecting files, not folders. My workaround is to create a new folder with the suffix `(number)` when another folder already exists.

It would be great if we had a complete node `fs` implementation to check files, folders, etc.

**Reply by tuxpeople** [reply]
Wouldn't it be possible to use `this.app.vault.adapter.exists(dir)` for this? I found that somewhere else, but can't confirm if it's really working or not.

**Reply by tuxpeople** [reply]
Still not 100% sure, but it looks like this is working:

```
if (!(await this.app.vault.adapter.exists(dir))) {
    await this.app.vault.createFolder(dir);
}
```

**Reply by AB1908** [reply]
You are correct.

**Reply by welpdx** [reply]
you can also you this to check if dir exists: 
```js 
if (!await this.app.vault.getAbstractFileByPath(dir)) {
    await this.app.vault.createFolder(dir);
}
```
It can also be used to get the t file of the folder.
